# **HOMEWORK 5** 

**VIGNESH PRABHAKAR**<br>
**USC ID:** : 4024890667


Mounted the google drive in colab workspace 

In [0]:
from google.colab import drive
drive.mount('/gdrive')

Drive already mounted at /gdrive; to attempt to forcibly remount, call drive.mount("/gdrive", force_remount=True).


# (1)(a)

Loaded the Anuran calls dataset from the drive

In [0]:
import pandas as pd

df= pd.read_csv('/gdrive/My Drive/Homeworks/hw5/Frogs_MFCCs.csv')
df.head()

MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  ...      Genus         Species  RecordID
0       1.0  0.152936 -0.105586  ...  Adenomera  AdenomeraAndre         1
1       1.0  0.171534 -0.098975  ...  Adenomera  AdenomeraAndre         1
2       1.0  0.152317 -0.082973  ...  Adenomera  AdenomeraAndre         1
3       1.0  0.224392  0.118985  ...  Adenomera  AdenomeraAndre         1
4       1.0  0.087817 -0.068345  ...  Adenomera  AdenomeraAndre         1

[5 rows x 26 columns]

Separated the features from the response labels . Here we have 3 response labels namely family ,genus and species . Each label has multiple classes in them . Thereby this is a multi class multi label problem .

In [0]:
x= df.iloc[:,0:22]
x

MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  ...  MFCCs_20  MFCCs_21  MFCCs_22
0          1.0  0.152936 -0.105586  ...  0.057684  0.118680  0.014038
1          1.0  0.171534 -0.098975  ...  0.020140  0.082263  0.029056
2          1.0  0.152317 -0.082973  ... -0.025083  0.099108  0.077162
3          1.0  0.224392  0.118985  ... -0.054766 -0.018691  0.023954
4          1.0  0.087817 -0.068345  ... -0.031346  0.108610  0.079244
...        ...       ...       ...  ...       ...       ...       ...
7190       1.0 -0.554504 -0.337717  ...  0.052449 -0.021860 -0.079860
7191       1.0 -0.517273 -0.370574  ...  0.046461 -0.015418 -0.101892
7192       1.0 -0.582557 -0.343237  ...  0.027834 -0.000531 -0.080425
7193       1.0 -0.519497 -0.307553  ...  0.041803 -0.027911 -0.096895
7194       1.0 -0.508833 -0.324106  ...  0.031560 -0.029355 -0.087910

[7195 rows x 22 columns]

In [0]:
Y=df[['Family','Genus','Species']]
Y

Family      Genus         Species
0     Leptodactylidae  Adenomera  AdenomeraAndre
1     Leptodactylidae  Adenomera  AdenomeraAndre
2     Leptodactylidae  Adenomera  AdenomeraAndre
3     Leptodactylidae  Adenomera  AdenomeraAndre
4     Leptodactylidae  Adenomera  AdenomeraAndre
...               ...        ...             ...
7190          Hylidae     Scinax     ScinaxRuber
7191          Hylidae     Scinax     ScinaxRuber
7192          Hylidae     Scinax     ScinaxRuber
7193          Hylidae     Scinax     ScinaxRuber
7194          Hylidae     Scinax     ScinaxRuber

[7195 rows x 3 columns]

Train test split . Randomly allocated 70% of the data as the training set and 30% of the data as the test set .

In [0]:
from sklearn.model_selection import train_test_split
train_x, test_x, train_y, test_y = train_test_split(x, Y, test_size=0.30, random_state=42)

Shape of training features , training responses ,test features and test responses

In [0]:
print(train_x.shape)
print(test_x.shape)
print(train_y.shape)
print(test_y.shape)

(5036, 22)
(2159, 22)
(5036, 3)
(2159, 3)


Extracted all response labels and stored each of them in separate variables so that they can be used easily to fit the classifiers that we will train .

In [0]:
label_1_test = test_y['Family']
label_2_test = test_y['Genus']
label_3_test = test_y['Species']

In [0]:
label_1_train = train_y['Family']
label_2_train = train_y['Genus']
label_3_train = train_y['Species']

# (1)(b)(i)

The Hamming loss is the fraction of labels that are incorrectly predicted. <br>

The Hamming score is the fraction of labels that are correctly predicted. <br>

An exact match occurs when the predicted value and actual value matches for all the 3 labels . <br>

Hamming loss is computed through the hamming_loss() function in sklearn.metrics and hamming score is computed by subtracting the hamming loss from 1 .





In [0]:
lbl1_score=list()
lbl2_score=list()
lbl3_score=list()


# (1)(b)(ii)

Computed the cross validation score on 3 separate classifiers . One that has family as train_y , second that has genus as train_y and third that has species as train_y and then selected the value of C and gamma corresponding to the maximum cross val score for each of these classifiers and then fit our gaussian kernelized support vector machine for different labels as train_y and used the C and gamma values that gave me the highest cross validation score for the classifier corresponding to that label .

In [0]:
from sklearn.model_selection import cross_val_score
from sklearn.svm import SVC
from sklearn import svm

C_vals=[0.001,0.01,0.1,1,10,100,1000]
g_vals=[0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1]
C_list=[]
g_list=[]
for g in g_vals:
  for k in C_vals:
    g_list.append(g)
    C_list.append(k)
    lbl1_score.append(cross_val_score(svm.SVC(C=k,kernel='rbf',gamma=g,decision_function_shape='ovr'),train_x,label_1_train,cv=10).mean())
    lbl2_score.append(cross_val_score(svm.SVC(C=k,kernel='rbf',gamma=g,decision_function_shape='ovr'),train_x,label_2_train,cv=10).mean())
    lbl3_score.append(cross_val_score(svm.SVC(C=k,kernel='rbf',gamma=g,decision_function_shape='ovr'),train_x,label_3_train,cv=10).mean())

C values and gamma values corresponding to their respective cross validation scores as a table with family label

In [0]:
C_gamma_selector_family = pd.DataFrame({"C-values":C_list,"Gamma values":g_list,"Family":lbl1_score})
C_gamma_selector_family    

C-values  Gamma values    Family
0      0.001           0.1  0.610207
1      0.010           0.1  0.620333
2      0.100           0.1  0.870943
3      1.000           0.1  0.943223
4     10.000           0.1  0.967841
..       ...           ...       ...
65     0.100           1.0  0.961292
66     1.000           1.0  0.986694
67    10.000           1.0  0.990665
68   100.000           1.0  0.992651
69  1000.000           1.0  0.992054

[70 rows x 3 columns]

Chose C =100 and gamma = 0.9 for family label

In [0]:
C_gamma_selector_family[C_gamma_selector_family['Family']==max(lbl1_score)]

C-values  Gamma values    Family
61     100.0           0.9  0.992651
68     100.0           1.0  0.992651

C values and gamma values corresponding to their respective cross validation scores as a table with Genus label

In [0]:
C_gamma_selector_genus = pd.DataFrame({"C-values":C_list,"Gamma values":g_list,"Genus":lbl2_score})
C_gamma_selector_genus 

C-values  Gamma values     Genus
0      0.001           0.1  0.575662
1      0.010           0.1  0.575662
2      0.100           0.1  0.825073
3      1.000           0.1  0.935479
4     10.000           0.1  0.974594
..       ...           ...       ...
65     0.100           1.0  0.935469
66     1.000           1.0  0.983317
67    10.000           1.0  0.989269
68   100.000           1.0  0.989869
69  1000.000           1.0  0.989869

[70 rows x 3 columns]

Chose C =100 and gamma = 1.0 for Genus label

In [0]:
C_gamma_selector_genus[C_gamma_selector_genus['Genus']==max(lbl2_score)]

C-values  Gamma values     Genus
68     100.0           1.0  0.989869
69    1000.0           1.0  0.989869

C values and gamma values corresponding to their respective cross validation scores as a table with Species label

In [0]:
C_gamma_selector_species = pd.DataFrame({"C-values":C_list,"Gamma values":g_list,"Species":lbl3_score})
C_gamma_selector_species

C-values  Gamma values   Species
0      0.001           0.1  0.485910
1      0.010           0.1  0.485910
2      0.100           0.1  0.839978
3      1.000           0.1  0.943214
4     10.000           0.1  0.978160
..       ...           ...       ...
65     0.100           1.0  0.939448
66     1.000           1.0  0.985691
67    10.000           1.0  0.989469
68   100.000           1.0  0.988470
69  1000.000           1.0  0.988470

[70 rows x 3 columns]

Chose C =10 and gamma = 1.0 for Species label

In [0]:
C_gamma_selector_species[C_gamma_selector_species['Species']==max(lbl3_score)]

C-values  Gamma values   Species
67      10.0           1.0  0.989469

Retraining the gaussian kernelized SVM for family label with chosen C and gamma and computing hamming loss and exact match score for the same .

Hamming loss :0.006947660954145438 <br>
Exact Match : 0.9930523390458545

In [0]:
from sklearn.metrics import hamming_loss
mdl=svm.SVC(C=100,gamma=0.9,decision_function_shape='ovr')
mdl=mdl.fit(train_x,label_1_train)
preds1=mdl.predict(test_x)
hm_loss_family=hamming_loss(label_1_test,preds1)
exact_match_family=mdl.score(test_x,label_1_test)
print(hm_loss_family)
print(exact_match_family)

0.006947660954145438
0.9930523390458545


Retraining the gaussian kernelized SVM for Genus label with chosen C and gamma and computing hamming loss and exact match score for the same .

Hamming loss :0.007874015748031496 <br>
Exact match : 0.9921259842519685

In [0]:
from sklearn.metrics import hamming_loss
mdl=svm.SVC(C=100,gamma=1.0,decision_function_shape='ovr')
mdl=mdl.fit(train_x,label_2_train)
preds2=mdl.predict(test_x)
hm_loss_genus=hamming_loss(label_2_test,preds2)
exact_match_genus=mdl.score(test_x,label_2_test)
print(hm_loss_genus)
print(exact_match_genus)

0.007874015748031496
0.9921259842519685


Retraining the gaussian kernelized SVM for Species label with chosen C and gamma and computing hamming loss and exact match score for the same .

Hamming loss :0.010653080129689671 <br>
Exact Match : 0.9893469198703103

In [0]:
from sklearn.metrics import hamming_loss
mdl=svm.SVC(C=10,gamma=1.0,decision_function_shape='ovr')
mdl=mdl.fit(train_x,label_3_train)
preds3=mdl.predict(test_x)
hm_loss_species=hamming_loss(label_3_test,preds3)
exact_match_species=mdl.score(test_x,label_3_test)
print(hm_loss_species)
print(exact_match_species)

0.010653080129689671
0.9893469198703103


Overall hamming loss and exact match computed <br>

The overall hamming loss is calculated by : <br>

sum(not_equal(test_y_overall, preds_overall))/test_y_overall_size

The overall exact match score is computed when the predicted value and actual value match for all the 3 labels in the same given row  

Overall hamming loss : 0.008491585610622202 <br>
Overall exact match : 0.9856415006947661


In [0]:
import numpy as np
preds_overall= np.concatenate((preds1.reshape(-1,1),preds2.reshape(-1,1),preds3.reshape(-1,1)),axis=1)
test_y_overall= np.concatenate((np.array(label_1_test).reshape(-1,1),np.array(label_2_test).reshape(-1,1),np.array(label_3_test).reshape(-1,1)),axis=1)

In [0]:
test_y_overall_size=float(test_y_overall.size)
total_hamming_loss = np.sum(np.not_equal(test_y_overall, preds_overall))/test_y_overall_size
print(total_hamming_loss)

0.008491585610622202


In [0]:
t=0
for i in range(0,len(preds_overall)):
  if(preds1[i]==label_1_test.iloc[i,] and preds2[i]==label_2_test.iloc[i,] and preds3[i]==label_3_test.iloc[i,]):
    t=t+1
t=t/len(preds_overall)
print(t)

0.9856415006947661


# (1)(b)(iii) 

Computed the cross validation score on 3 separate L1 penalized linear SVM classifiers . One that has family as train_y , second that has genus as train_y and third that has species as train_y and then selected the value of C corresponding to the maximum cross val score for each of these classifiers and then fit our linear support vector classifier for different labels as train_y and used the C values that gave me the highest cross validation score for the classifier corresponding to that label .

In [0]:
from sklearn.svm import LinearSVC
lbl1_l1svm_score=list()
lbl2_l1svm_score=list()
lbl3_l1svm_score=list()




Standardized the features due to the presence of the L-1 penalty .

In [0]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
train_x = sc.fit_transform(train_x)
test_x = sc.transform(test_x)

In [0]:
C_vals=[0.001,0.01,0.1,1,10,100,1000]
C1_list=[]

for k in C_vals:
    C1_list.append(k)
    mdl = LinearSVC(C=k,penalty='l1',dual=False)
    lbl1_l1svm_score.append(cross_val_score(mdl,train_x,label_1_train,cv=10).mean())
    lbl2_l1svm_score.append(cross_val_score(mdl,train_x,label_2_train,cv=10).mean())
    lbl3_l1svm_score.append(cross_val_score(mdl,train_x,label_3_train,cv=10).mean())

/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


C values corresponding to their respective cross validation scores as a table with family label

In [0]:
C_l1_selector_family = pd.DataFrame({"C-values":C1_list,"Family":lbl1_l1svm_score})
C_l1_selector_family    

C-values    Family
0     0.001  0.852855
1     0.010  0.929516
2     0.100  0.938056
3     1.000  0.940632
4    10.000  0.940433
5   100.000  0.940235
6  1000.000  0.940235

Chose C =1 for Family label

In [0]:
C_l1_selector_family[C_l1_selector_family['Family']==max(lbl1_l1svm_score)]

C-values    Family
3       1.0  0.940632

C values corresponding to their respective cross validation scores as a table with Genus label

In [0]:
C_l1_selector_genus = pd.DataFrame({"C-values":C1_list,"Genus":lbl2_l1svm_score})
C_l1_selector_genus    

C-values     Genus
0     0.001  0.823683
1     0.010  0.913818
2     0.100  0.942619
3     1.000  0.950956
4    10.000  0.952751
5   100.000  0.952948
6  1000.000  0.952948

Chose C =100 for Genus label

In [0]:
C_l1_selector_genus[C_l1_selector_genus['Genus']==max(lbl2_l1svm_score)]

C-values     Genus
5     100.0  0.952948
6    1000.0  0.952948

C values corresponding to their respective cross validation scores as a table with Species label

In [0]:
C_l1_selector_species = pd.DataFrame({"C-values":C1_list,"Species":lbl3_l1svm_score})
C_l1_selector_species

C-values   Species
0     0.001  0.786936
1     0.010  0.915013
2     0.100  0.949755
3     1.000  0.960684
4    10.000  0.960483
5   100.000  0.960285
6  1000.000  0.960285

Chose C =1 for Genus label

In [0]:
C_l1_selector_species[C_l1_selector_species['Species']==max(lbl3_l1svm_score)]

C-values   Species
3       1.0  0.960684

Retraining the L-1 penalized linear SVC for family label with chosen C and computing hamming loss and exact match score for the same .

Hamming loss :0.07132931912922649 <br>
Exact Match : 0.9286706808707735

In [0]:
from sklearn.metrics import hamming_loss
mdl = LinearSVC(C=1,penalty='l1',dual=False)
mdl=mdl.fit(train_x,label_1_train)
preds1=mdl.predict(test_x)
hm_loss_family=hamming_loss(label_1_test,preds1)
exact_match_family=mdl.score(test_x,label_1_test)
print(hm_loss_family)
print(exact_match_family)

0.07132931912922649
0.9286706808707735


/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


Retraining the L-1 penalized linear SVC for Genus label with chosen C and computing hamming loss and exact match score for the same .

Hamming loss :0.058360352014821676 <br>
Exact Match : 0.9416396479851783

In [0]:
from sklearn.metrics import hamming_loss
mdl = LinearSVC(C=100,penalty='l1',dual=False)
mdl=mdl.fit(train_x,label_2_train)
preds2=mdl.predict(test_x)
hm_loss_genus=hamming_loss(label_2_test,preds2)
exact_match_genus=mdl.score(test_x,label_2_test)
print(hm_loss_genus)
print(exact_match_genus)

0.058360352014821676
0.9416396479851783


/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


Retraining the L-1 penalized linear SVC for Species label with chosen C and computing hamming loss and exact match score for the same .

Hamming loss :0.041222788327929596 <br>
Exact Match : 0.9587772116720704

In [0]:
from sklearn.metrics import hamming_loss
mdl = LinearSVC(C=1,penalty='l1',dual=False)
mdl=mdl.fit(train_x,label_3_train)
preds3=mdl.predict(test_x)
hm_loss_species=hamming_loss(label_3_test,preds3)
exact_match_species=mdl.score(test_x,label_3_test)
print(hm_loss_species)
print(exact_match_species)

0.041222788327929596
0.9587772116720704


/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


Overall hamming loss and exact match computed <br>

The overall hamming loss is calculated by : <br>

sum(not_equal(test_y_overall, preds_overall))/test_y_overall_size

The overall exact match score is computed when the predicted value and actual value match for all the 3 labels in the same given row  

Overall hamming loss : 0.05697081982399259 <br>
Overall exact match : 0.9133858267716536

In [0]:
import numpy as np
preds_overall_l1= np.concatenate((preds1.reshape(-1,1),preds2.reshape(-1,1),preds3.reshape(-1,1)),axis=1)
test_y_overall_l1= np.concatenate((np.array(label_1_test).reshape(-1,1),np.array(label_2_test).reshape(-1,1),np.array(label_3_test).reshape(-1,1)),axis=1)

In [0]:
test_y_overall_l1_size=float(test_y_overall_l1.size)
total_hamming_loss_l1 = np.sum(np.not_equal(test_y_overall_l1, preds_overall_l1))/test_y_overall_l1_size
print(total_hamming_loss_l1)

0.05697081982399259


In [0]:
t=0
for i in range(0,len(preds_overall_l1)):
  if(preds1[i]==label_1_test.iloc[i,] and preds2[i]==label_2_test.iloc[i,] and preds3[i]==label_3_test.iloc[i,]):
    t=t+1
t=t/len(preds_overall_l1)
print(t)

0.9133858267716536


# (1)(b)(iv)

Computed the cross validation score on 3 separate L1 penalized linear SVM classifiers . One that has family as train_y , second that has genus as train_y and third that has species as train_y and then selected the value of C corresponding to the maximum cross val score for each of these classifiers and then fit our linear support vector classifier for different labels as train_y and used the C values that gave me the highest cross validation score for the classifier corresponding to that label .

Remove the class imbalance using SMOTE ( synthetic minority oversampling technique) and then use the data sampled using SMOTE in order to fit the L1 penalized linear SVC classifier

In [0]:
from imblearn.over_sampling import SMOTE 
sampling=SMOTE(kind='svm',k_neighbors=5)



In [0]:
smote_tx_lbl1,smote_ty_lbl1=sampling.fit_sample(train_x,label_1_train)
smote_tx_lbl2,smote_ty_lbl2=sampling.fit_sample(train_x,label_2_train)
smote_tx_lbl3,smote_ty_lbl3=sampling.fit_sample(train_x,label_3_train)

In [0]:
lbl1_smsvm_score=list()
lbl2_smsvm_score=list()
lbl3_smsvm_score=list()

In [0]:
C_vals=[0.001,0.01,0.1,1,10,100,1000]
C2_list=[]

for k in C_vals:
    C2_list.append(k)
    mdl = LinearSVC(C=k,penalty='l1',dual=False)
    lbl1_smsvm_score.append(cross_val_score(mdl,smote_tx_lbl1,smote_ty_lbl1,cv=10).mean())
    lbl2_smsvm_score.append(cross_val_score(mdl,smote_tx_lbl2,smote_ty_lbl2,cv=10).mean())
    lbl3_smsvm_score.append(cross_val_score(mdl,smote_tx_lbl3,smote_ty_lbl3,cv=10).mean())

/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [0]:
C_sm_selector_family = pd.DataFrame({"C-values":C2_list,"Family":lbl1_smsvm_score})

Chose C =10 for Family label

In [0]:
C_sm_selector_family[C_sm_selector_family['Family']==max(lbl1_smsvm_score)]

C-values    Family
4      10.0  0.894783
5     100.0  0.894783
6    1000.0  0.894783

In [0]:
C_sm_selector_genus = pd.DataFrame({"C-values":C2_list,"Genus":lbl2_smsvm_score})

Chose C =10 for Genus label

In [0]:
C_sm_selector_genus[C_sm_selector_genus['Genus']==max(lbl2_smsvm_score)]

C-values     Genus
4      10.0  0.909306

In [0]:
C_sm_selector_species = pd.DataFrame({"C-values":C2_list,"Species":lbl3_smsvm_score})

Chose C =10 for Species label

In [0]:
C_sm_selector_species[C_sm_selector_species['Species']==max(lbl3_smsvm_score)]

C-values  Species
4      10.0  0.89872

Retraining the L-1 penalized linear SVC for Family label with chosen C and computing hamming loss and exact match score for the same .

Hamming loss :0.11718388142658638 <br>
Exact Match : 0.8828161185734136

In [0]:
from sklearn.metrics import hamming_loss
mdl = LinearSVC(C=10,penalty='l1',dual=False)
mdl=mdl.fit(smote_tx_lbl1,smote_ty_lbl1)
preds1=mdl.predict(test_x)
hm_loss_family=hamming_loss(label_1_test,preds1)
exact_match_family=mdl.score(test_x,label_1_test)
print(hm_loss_family)
print(exact_match_family)

0.11718388142658638
0.8828161185734136


/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


Retraining the L-1 penalized linear SVC for Genus label with chosen C and computing hamming loss and exact match score for the same .

Hamming loss :0.14173228346456693 <br>
Exact Match : 0.8582677165354331

In [0]:
from sklearn.metrics import hamming_loss
mdl = LinearSVC(C=10,penalty='l1',dual=False)
mdl=mdl.fit(smote_tx_lbl2,smote_ty_lbl2)
preds2=mdl.predict(test_x)
hm_loss_genus=hamming_loss(label_2_test,preds2)
exact_match_genus=mdl.score(test_x,label_2_test)
print(hm_loss_genus)
print(exact_match_genus)

0.14173228346456693
0.8582677165354331


/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


Retraining the L-1 penalized linear SVC for Species label with chosen C and computing hamming loss and exact match score for the same .

Hamming loss :0.06947660954145438 <br> 
Exact Match : 0.9305233904585456

In [0]:
from sklearn.metrics import hamming_loss
mdl = LinearSVC(C=10,penalty='l1',dual=False)
mdl=mdl.fit(smote_tx_lbl3,smote_ty_lbl3)
preds3=mdl.predict(test_x)
hm_loss_species=hamming_loss(label_3_test,preds3)
exact_match_species=mdl.score(test_x,label_3_test)
print(hm_loss_species)
print(exact_match_species)

0.06947660954145438
0.9305233904585456


/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


Overall hamming loss and exact match computed <br>

The overall hamming loss is calculated by : <br>

sum(not_equal(test_y_overall, preds_overall))/test_y_overall_size

The overall exact match score is computed when the predicted value and actual value match for all the 3 labels in the same given row  

Overall hamming loss : 0.10946425814420256 <br>
Overall exact match : 0.8012968967114404

In [0]:
import numpy as np
preds_overall_sm= np.concatenate((preds1.reshape(-1,1),preds2.reshape(-1,1),preds3.reshape(-1,1)),axis=1)
test_y_overall_sm= np.concatenate((np.array(label_1_test).reshape(-1,1),np.array(label_2_test).reshape(-1,1),np.array(label_3_test).reshape(-1,1)),axis=1)

In [0]:
test_y_overall_sm_size=float(test_y_overall_sm.size)
total_hamming_loss_sm = np.sum(np.not_equal(test_y_overall_sm, preds_overall_sm))/test_y_overall_sm_size
print(total_hamming_loss_sm)

0.10946425814420256


In [0]:
t=0
for i in range(0,len(preds_overall_sm)):
  if(preds1[i]==label_1_test.iloc[i,] and preds2[i]==label_2_test.iloc[i,] and preds3[i]==label_3_test.iloc[i,]):
    t=t+1
t=t/len(preds_overall_sm)
print(t)

0.8012968967114404


**Conclusions about the classifiers trained** <br>


             Hamming Loss   Exact Match Score 
Gaussian :  	  0.008491	      0.985641 <br>

---


l1 penalty SVM :	0.056970	     0.913385 <br>

---


resampled l1 SVM	0.109464	    0.801296<br>


When our data points aren't linearly separable in 2 dimensions we want to be able to transform them to a higher dimension for separability by kernelization . This is successfully achieved using a RBF kernelized SVM and we obtain the highest exact match score and lowest hamming loss when compared to the other two methods . 

This is mainly because the data points in the anuran calls dataset are not exactly linearly separable . Thus kernelized SVM gives us the better performance amongst these 3 methods .

Mounted google drive into colab workspace

In [0]:
from google.colab import drive
drive.mount('/gdrive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /gdrive


Created a dataframe by loading the anuran calls dataset from the drive

In [0]:
import pandas as pd
df=pd.read_csv('/gdrive/My Drive/Homeworks/hw5/Frogs_MFCCs.csv',usecols=range(0,25))
df.head()

MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  ...           Family      Genus         Species
0       1.0  0.152936 -0.105586  ...  Leptodactylidae  Adenomera  AdenomeraAndre
1       1.0  0.171534 -0.098975  ...  Leptodactylidae  Adenomera  AdenomeraAndre
2       1.0  0.152317 -0.082973  ...  Leptodactylidae  Adenomera  AdenomeraAndre
3       1.0  0.224392  0.118985  ...  Leptodactylidae  Adenomera  AdenomeraAndre
4       1.0  0.087817 -0.068345  ...  Leptodactylidae  Adenomera  AdenomeraAndre

[5 rows x 25 columns]

In [0]:
overall_hamming_loss=list()
overall_hamming_score=list()

Segregated the features from the labels 

In [0]:
x = df.iloc[:,0:22]
y = df.iloc[:,22:25]


In [0]:
x.head()

MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  ...  MFCCs_20  MFCCs_21  MFCCs_22
0       1.0  0.152936 -0.105586  ...  0.057684  0.118680  0.014038
1       1.0  0.171534 -0.098975  ...  0.020140  0.082263  0.029056
2       1.0  0.152317 -0.082973  ... -0.025083  0.099108  0.077162
3       1.0  0.224392  0.118985  ... -0.054766 -0.018691  0.023954
4       1.0  0.087817 -0.068345  ... -0.031346  0.108610  0.079244

[5 rows x 22 columns]

In [0]:
y.head()

Family      Genus         Species
0  Leptodactylidae  Adenomera  AdenomeraAndre
1  Leptodactylidae  Adenomera  AdenomeraAndre
2  Leptodactylidae  Adenomera  AdenomeraAndre
3  Leptodactylidae  Adenomera  AdenomeraAndre
4  Leptodactylidae  Adenomera  AdenomeraAndre

# (2)(a) , (2)(b) , (2)(c)

In [0]:
family_majority={}
genus_majority={}
species_majority={}
for  p in range(1,51):
  family_majority[p]=list()
  genus_majority[p]= list()
  species_majority[p]=list()

*   We have 2 nested for loops for monte carlo and kmeans.   

*   Within the outer for loop we calculate the mean and standard deviation of the hamming distances

*   To do this, we are predicting labels within the for loop and calculating hamming distances by comparing true labels and labels assigned by clusters

*   So the inner for loop, predicts labels and calculates silhouette score accordingly and stores it.


*   The outer one then checks for the max silhouette score and the k value that corresponds to it and performs a prediction for the label again.

*   With this label we calculate the hamming distances and store it.

*   Then we calculate the mean and standard deviations of the hamming distances

*Some terminologies and their meanings :*



1.   Hamming distance : Hamming distance is the no. of mismatches between 2 vectors <br>

2.   Hamming score : Hamming score is the no. of matches between 2 vectors <br>
     It can also be represented as  (1 - hamming loss)

3.   Hamming loss : Average of the hamming distances is hamming loss <br>














In [0]:
from sklearn import metrics
from sklearn.cluster import KMeans
from sklearn.metrics import hamming_loss
import random
import numpy as np


for i in range(1,51):
  silhouette_analysis = pd.DataFrame(columns=['K-value', 'Silhouette-score'])
  for k in range(2,50):
    mdl = KMeans(n_clusters=k,init='k-means++',random_state=random.randint(0, 800)).fit(x)
    lbl = mdl.labels_
    silhouette_analysis = silhouette_analysis.append({'K-value': k, 'Silhouette-score':metrics.silhouette_score(x, lbl)}, ignore_index=True)
  chosen_k = silhouette_analysis['K-value'][silhouette_analysis['Silhouette-score']==max(silhouette_analysis['Silhouette-score'])]
  chosen_k=int(chosen_k)
  k_means_chosenk = KMeans(n_clusters=chosen_k, random_state=random.randint(0, 800)).fit(x)
  cls_lbl = k_means_chosenk.labels_

  cls = pd.concat([df,pd.DataFrame({'multi_class_labels':cls_lbl.tolist()})],axis = 1)
  majority={}
  for k_val in range(0,chosen_k):
    majority[k_val]=list()

  for k in range(0,chosen_k):
    lbl_name = cls[cls['multi_class_labels']==k]
    majority[k].append(lbl_name['Family'].value_counts().index[0])
    family_majority[i].append(lbl_name['Family'].value_counts().index[0])
    majority[k].append(lbl_name['Genus'].value_counts().index[0])
    genus_majority[i].append(lbl_name['Genus'].value_counts().index[0])
    majority[k].append(lbl_name['Species'].value_counts().index[0])
    species_majority[i].append(lbl_name['Species'].value_counts().index[0])

  cls['family_preds'] = ''
  cls['genus_preds'] = ''
  cls['species_preds'] = ''

  for k in range(0,chosen_k):
    cls['family_preds'] = np.where(cls['multi_class_labels']==k,majority[k][0],cls['family_preds'])
    cls['genus_preds'] = np.where(cls['multi_class_labels']==k,majority[k][1],cls['genus_preds'])
    cls['species_preds'] = np.where(cls['multi_class_labels']==k,majority[k][2],cls['species_preds'])

    
  hamming_family_loss=hamming_loss(cls['Family'],cls['family_preds'])
  hamming_genus_loss=hamming_loss(cls['Genus'],cls['genus_preds'])
  hamming_species_loss=hamming_loss(cls['Species'],cls['species_preds'])

  
  total_hamming_loss= (hamming_family_loss+hamming_genus_loss+hamming_species_loss)/3

  overall_hamming_loss.append(total_hamming_loss)
  overall_hamming_score.append(1-total_hamming_loss)



In [0]:
print("Mean of the hamming distances : ",np.mean(overall_hamming_loss))
print("Std deviation of the hamming distances : ",np.std(overall_hamming_loss))


Mean of the hamming distances :  0.22521380588371553
Std deviation of the hamming distances :  0.00686636398271826


In [0]:
m =list(zip(list(family_majority.values()),list(genus_majority.values()),list(species_majority.values())))

In [0]:
majority=dict()
for j in range(0,50):
    majority[j]=list(zip(m[j][0],m[j][1],m[j][2]))

Table summarizing the Optimal_K chosen using silhouette score , Hamming_score, hamming_loss and majority_label_triplet that has been achieved over 50 iterations of the monte carlo simulation 

In [0]:
Monte_carlo_simulation=pd.DataFrame({"Optimal K":chosen_k,"Hamming Score":overall_hamming_score,"Hamming Loss" : overall_hamming_loss,"Majority label triplet":list(majority.values())})
Monte_carlo_simulation

Optimal K  ...                             Majority label triplet
0           4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
1           4  ...  [(Hylidae, Hypsiboas, HypsiboasCordobae), (Lep...
2           4  ...  [(Hylidae, Hypsiboas, HypsiboasCinerascens), (...
3           4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
4           4  ...  [(Hylidae, Hypsiboas, HypsiboasCordobae), (Lep...
5           4  ...  [(Hylidae, Hypsiboas, HypsiboasCinerascens), (...
6           4  ...  [(Hylidae, Hypsiboas, HypsiboasCinerascens), (...
7           4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
8           4  ...  [(Leptodactylidae, Adenomera, AdenomeraAndre),...
9           4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
10          4  ...  [(Hylidae, Hypsiboas, HypsiboasCordobae), (Lep...
11          4  ...  [(Hylidae, Hypsiboas, HypsiboasCordobae), (Hyl...
12          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
13          4  ...  [(Hylidae, Hypsiboas, HypsiboasCinerascens), (...
14          4  ...  [(Dendrobatidae, Ameerega, Ameeregatrivittata)...
15          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
16          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
17          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
18          4  ...  [(Dendrobatidae, Ameerega, Ameeregatrivittata)...
19          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
20          4  ...  [(Dendrobatidae, Ameerega, Ameeregatrivittata)...
21          4  ...  [(Hylidae, Hypsiboas, HypsiboasCinerascens), (...
22          4  ...  [(Dendrobatidae, Ameerega, Ameeregatrivittata)...
23          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
24          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
25          4  ...  [(Hylidae, Hypsiboas, HypsiboasCordobae), (Lep...
26          4  ...  [(Hylidae, Hypsiboas, HypsiboasCordobae), (Den...
27          4  ...  [(Hylidae, Hypsiboas, HypsiboasCinerascens), (...
28          4  ...  [(Hylidae, Hypsiboas, HypsiboasCinerascens), (...
29          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
30          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
31          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
32          4  ...  [(Hylidae, Hypsiboas, HypsiboasCinerascens), (...
33          4  ...  [(Hylidae, Hypsiboas, HypsiboasCordobae), (Lep...
34          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
35          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
36          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
37          4  ...  [(Hylidae, Hypsiboas, HypsiboasCinerascens), (...
38          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
39          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
40          4  ...  [(Dendrobatidae, Ameerega, Ameeregatrivittata)...
41          4  ...  [(Hylidae, Hypsiboas, HypsiboasCordobae), (Lep...
42          4  ...  [(Hylidae, Hypsiboas, HypsiboasCinerascens), (...
43          4  ...  [(Hylidae, Hypsiboas, HypsiboasCordobae), (Den...
44          4  ...  [(Dendrobatidae, Ameerega, Ameeregatrivittata)...
45          4  ...  [(Hylidae, Hypsiboas, HypsiboasCordobae), (Lep...
46          4  ...  [(Leptodactylidae, Adenomera, AdenomeraHylaeda...
47          4  ...  [(Dendrobatidae, Ameerega, Ameeregatrivittata)...
48          4  ...  [(Hylidae, Hypsiboas, HypsiboasCinerascens), (...
49          4  ...  [(Hylidae, Hypsiboas, HypsiboasCordobae), (Lep...

[50 rows x 4 columns]

#(3)  ISLR 10.7.2 

**For the ISLR portions of this homework , kindly refer to the PDF document attached along with this in the .zip file that has been submitted**